### PR TITLE
[chart-fees-paid]: in and out flags alias support

### PR DESCRIPTION
**Pull request by niteshbalusu11**
Added support for entering alias/pubkey for in and out flags for `chart-fees-paid` command.

https://github.com/alexbosworth/balanceofsatoshis/issues/272

Files changed:
bos
routing/get_fees_paid.js

### CODE REVIEW

File: routing/get_fees_paid.js [hunk 1]
```diff
@@ -90,6 +93,56 @@ module.exports = (args, cbk) => {
...
+
+        return asyncMap(args.lnds, (lnd, cbk) => {
+          return findKey({lnd, query: args.in}, cbk);
+        },
+        (err, res) => {
+          if (!!err) {
+            return cbk(err);
+          }
+
+          if (res.length > 1) {
```
Comment (alexbosworth) [inline routing/get_fees_paid.js:111]: wouldn't you want to unique the public key here, and see if that is more than one? Maybe they both have the same peer, so same match

File: routing/get_fees_paid.js [hunk 2]
```diff
@@ -90,6 +93,56 @@ module.exports = (args, cbk) => {
...
+
+        return asyncMap(args.lnds, (lnd, cbk) => {
+          return findKey({lnd, query: args.out}, cbk);
+        },
+        (err, res) => {
+          if (!!err) {
+            return cbk(err);
+          }
+
+          if (res.length > 1) {
```
Comment (alexbosworth) [inline routing/get_fees_paid.js:136]: same idea here

File: routing/get_fees_paid.js [hunk 3]
```diff
@@ -90,6 +93,56 @@ module.exports = (args, cbk) => {
...
+            return cbk([400, 'MultipleMatchesForOutQuery']);
```
Comment (alexbosworth) [inline routing/get_fees_paid.js:137]: Find key may return an empty object, like {}, so need to make sure it errors if there is no match on either

**Review by alexbosworth — CHANGES_REQUESTED** [review]

**Comment by niteshbalusu11** [comment]
> wouldn't you want to unique the public key here, and see if that is more than one? Maybe they both have the same peer, so same match 

Unique pubkey as in if multiple saved nodes are connected to the same peer?

**Comment by alexbosworth** [comment]
> > wouldn't you want to unique the public key here, and see if that is more than one? Maybe they both have the same peer, so same match
> 
> Unique pubkey as in if multiple saved nodes are connected to the same peer?

Yeah same peer node on multiple saved nodes

**Comment by niteshbalusu11** [comment]
Should be fixed now I think.

**Comment by niteshbalusu11** [comment]
> I think the error should be propagated here right? in case there are multiple matches it will give you a hint in the error

 ```         if (!allEqual(result)) {
            return cbk([400, 'MultipleMatchesForOutQuery']);
          }
```

I think this part does that, or you mean something else?

**Comment by alexbosworth** [comment]
> > I think the error should be propagated here right? in case there are multiple matches it will give you a hint in the error
> 
> ```
> if (!allEqual(result)) {
>      return cbk([400, 'MultipleMatchesForOutQuery']);
>  }
> ```
> 
> I think this part does that, or you mean something else?

yeah like on one node it has multiple matches

**Comment by niteshbalusu11** [comment]
> > > I think the error should be propagated here right? in case there are multiple matches it will give you a hint in the error
> > 
> > 
> > ```
> > if (!allEqual(result)) {
> >      return cbk([400, 'MultipleMatchesForOutQuery']);
> >  }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think this part does that, or you mean something else?
> 
> yeah like on one node it has multiple matches

Ok yeah, I see what you mean, I think the error should be displayed but not error out and continue checking?

Like an `AmbiguousAliasSpecified` error can happen on the first saved node but it might not occur on the 2nd saved node? Or if peer is not found on the 1st node but found on the 2nd node.

**Comment by alexbosworth** [comment]
> > > > I think the error should be propagated here right? in case there are multiple matches it will give you a hint in the error
> > > 
> > > 
> > > ```
> > > if (!allEqual(result)) {
> > >      return cbk([400, 'MultipleMatchesForOutQuery']);
> > >  }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I think this part does that, or you mean something else?
> > 
> > 
> > yeah like on one node it has multiple matches
> 
> Ok yeah, I see what you mean, I think the error should be displayed but not error out and continue checking?
> 
> Like an `AmbiguousAliasSpecified` error can happen on the first saved node but it might not occur on the 2nd saved node? Or if peer is not found on the 1st node but found on the 2nd node.

I think it should be an error if on any node and halt execution yeah

**Comment by niteshbalusu11** [comment]
> > > > > I think the error should be propagated here right? in case there are multiple matches it will give you a hint in the error
> > > > 
> > > > 
> > > > ```
> > > > if (!allEqual(result)) {
> > > >      return cbk([400, 'MultipleMatchesForOutQuery']);
> > > >  }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I think this part does that, or you mean something else?
> > > 
> > > 
> > > yeah like on one node it has multiple matches
> > 
> > 
> > Ok yeah, I see what you mean, I think the error should be displayed but not error out and continue checking?
> > Like an `AmbiguousAliasSpecified` error can happen on the first saved node but it might not occur on the 2nd saved node? Or if peer is not found on the 1st node but found on the 2nd node.
> 
> I think it should be an error if on any node and halt execution yeah

if `findKey` doesn't find something on saved node 1 and finds on 2, it will still error out for node 1 though and not check 2 at all, can we distinguish those errors?

**Comment by alexbosworth** [comment]
> if `findKey` doesn't find something on saved node 1 and finds on 2, it will still error out for node 1 though and not check 2 at all, can we distinguish those errors?

good point I wasn't thinking about that doh

this is tricky, we could look for `AmbiguousAliasSpecified` and only re-throw on that

or look for `FailedToFindPeerAliasMatch` and ignore that if the other node has the match

**Comment by niteshbalusu11** [comment]
> > if `findKey` doesn't find something on saved node 1 and finds on 2, it will still error out for node 1 though and not check 2 at all, can we distinguish those errors?
> 
> 
> 
> good point I wasn't thinking about that doh
> 
> 
> 
> this is tricky, we could look for `AmbiguousAliasSpecified` and only re-throw on that
> 
> 
> 
> or look for `FailedToFindPeerAliasMatch` and ignore that if the other node has the match

Ok, cool.

**Comment by niteshbalusu11** [comment]
You can check now.

**Comment by niteshbalusu11** [comment]
Hmm, what if it finds `--in` from one node and `--out` from another one?

**Comment by alexbosworth** [comment]
> Hmm, what if it finds `--in` from one node and `--out` from another one?

Yeah that is a good issue, in the node alias title, could add the first 8 chars of the pubkey to clarify the node identity that was selected

**Pull request merged**